src/parser/func_type.cpp

```diff
@@ -0,0 +1,155 @@
+  {0, {PT_TYPE_MAYBE}, {}},
+  {0, {PT_TYPE_NULL}, {}},
+  {0, {PT_TYPE_NA}, {}},
+  {0, {PT_TYPE_SET}, {}},
+  {0, {PT_TYPE_MULTISET}, {}},
+  {0, {PT_TYPE_SEQUENCE}, {}},
+};
+
+std::vector<func_signature> func_signature::double_r_any = {//original code doesn't check arguments!!!
+  {PT_TYPE_DOUBLE, {}, {PT_GENERIC_TYPE_ANY}},
+};
+
+std::vector<func_signature> func_signature::ntile = {//why original code cast args to double instead int???
+  {PT_TYPE_INTEGER, {PT_GENERIC_TYPE_DISCRETE_NUMBER}, {}},
+  {PT_TYPE_INTEGER, {PT_TYPE_MAYBE}, {}},
+  {PT_TYPE_INTEGER, {PT_TYPE_NULL}, {}},
+};
+
+std::vector<func_signature> func_signature::median = {
+  {PT_TYPE_DOUBLE, {PT_GENERIC_TYPE_NUMBER}, {}},
+  {0,  {PT_GENERIC_TYPE_DATETIME}, {}},
+};
+
+std::vector<func_signature> func_signature::type0_nr_or_str = {
+  {0, {PT_GENERIC_TYPE_NUMBER}, {}},
+  {0, {PT_GENERIC_TYPE_DATETIME}, {}},
+  {0, {PT_GENERIC_TYPE_STRING}, {}},
+  {0, {PT_GENERIC_TYPE_BIT}, {}},
+  {0, {PT_TYPE_ENUMERATION}, {}},
+  {0, {PT_TYPE_MAYBE}, {}},
+  {0, {PT_TYPE_NULL}, {}},
+  {0, {PT_TYPE_NA}, {}},
+};
+
+std::vector<func_signature> func_signature::type0_nr_or_str_discrete = {
+  {0, {PT_GENERIC_TYPE_NUMBER, PT_GENERIC_TYPE_DISCRETE_NUMBER}, {}},
+  {0, {PT_GENERIC_TYPE_DATETIME, PT_GENERIC_TYPE_DISCRETE_NUMBER}, {}},
+  {0, {PT_GENERIC_TYPE_STRING, PT_GENERIC_TYPE_DISCRETE_NUMBER}, {}},
+  {0, {PT_TYPE_ENUMERATION, PT_GENERIC_TYPE_DISCRETE_NUMBER}, {}},
+  {0, {PT_TYPE_MAYBE, PT_GENERIC_TYPE_DISCRETE_NUMBER}, {}},
+  {0, {PT_TYPE_NULL, PT_GENERIC_TYPE_DISCRETE_NUMBER}, {}},
+  {0, {PT_TYPE_NA, PT_GENERIC_TYPE_DISCRETE_NUMBER}, {}},
+};
+
+std::vector<func_signature> func_signature::group_concat = {
+//{PT_TYPE_VARCHAR, {PT_GENERIC_TYPE_STRING}, {}},
+  {PT_TYPE_VARCHAR , {PT_GENERIC_TYPE_CHAR  , PT_GENERIC_TYPE_CHAR  }, {}},
+  {PT_TYPE_VARNCHAR, {PT_GENERIC_TYPE_NCHAR , PT_GENERIC_TYPE_NCHAR }, {}},
+  {PT_TYPE_VARBIT  , {PT_GENERIC_TYPE_BIT   , PT_GENERIC_TYPE_BIT   }, {}},
+  {PT_TYPE_VARCHAR , {PT_GENERIC_TYPE_NUMBER, PT_GENERIC_TYPE_STRING}, {}},
+};
+
+std::vector<func_signature> func_signature::lead_lag = {//original code doesn't do anything!!!
+  {0, {PT_GENERIC_TYPE_NUMBER}, {}},
+  {0, {PT_GENERIC_TYPE_STRING}, {}},
+  {0, {PT_GENERIC_TYPE_DATETIME}, {}},
+  {0, {PT_GENERIC_TYPE_BIT}, {}},
+};
+
+std::vector<func_signature> func_signature::elt = {
+  {1           , {PT_GENERIC_TYPE_DISCRETE_NUMBER, PT_TYPE_VARCHAR      }, {1}},
+  {1           , {PT_GENERIC_TYPE_DISCRETE_NUMBER, PT_GENERIC_TYPE_NCHAR}, {1}},
+  {PT_TYPE_NULL, {PT_TYPE_NULL                                          }, {/*PT_TYPE_ANY*/}},
+  {PT_TYPE_NULL, {PT_TYPE_INTEGER                                       }, {}},
+};
+
+std::vector<func_signature> func_signature::insert = {
+  {PT_TYPE_VARCHAR , {PT_GENERIC_TYPE_CHAR , PT_TYPE_INTEGER, PT_TYPE_INTEGER, PT_GENERIC_TYPE_CHAR  }, {}},
+  {PT_TYPE_VARCHAR , {PT_GENERIC_TYPE_CHAR , PT_TYPE_INTEGER, PT_TYPE_INTEGER, PT_GENERIC_TYPE_NCHAR }, {}},
```

Comment:
We should not support mixing CHAR and NATIONAL CHAR.

Reply:
signatures can have 2 meanings:
1. to try casting to a specified type
2. to prevent casting and delegate the checking later on (after type checking)

I have no other way to explicitly specify that a specific signature is not allowed and prevent casting to another type from a (score) matched signature

There is a solution but Remzi disagreed: 
- use bitmasks for pt_type_enum values
- use bitmasks to describe allowed and forbidden signatures
- existing code needs no change, it uses the symbolic constants
- generic types will be defined also as bitmasks and the relation with pt_type_enum will be explicitly stated using C++ language features
- complex tests like:
    if(x==PT_TYPE_1 || x==PT_TYPE_2 || ...) 
will become simple tests like: 
    if(x & bitmask)
meaning less code and more performance

Reply:
Please consider using a score-based signature matching algorithm similar to expressions code.
This should allow to avoid including combinations of arguments of not allowed signatures.

Reply:
The existing functions used in expression code (pt_are_equivalent_types(), pt_get_equivalent_type()) are not enough for what we want to achieve for functions and that's why I introduced a new one that tells if an argument type could be casted to another type with the possibility to obtain something valid
e.g. a string could be casted to a valid number in some cases: '1234' -> 1234 (similar for date and other types)
This new function together with signatures containing only the allowed types, seems to be enough